### PR TITLE
Rewrite internals first batch

### PR DESCRIPTION
**Pull request by banciur**
First batch of rewrite of library.
Refactored `signTransaction` and `getAccounts` functions to use async functions
Rewritten `getMultipleAccounts` - change in method specification 
Added constrains on derivation path for getMultipleAccounts function
First unit tests

### CODE REVIEW

File: src/LedgerWallet.js [hunk 1]
```diff
@@ -158,109 +148,139 @@ class LedgerWallet {
...
-      const path = `${derivationPath}${i}`;
-      const address = await eth.getAddress_async(
-        path,
-        this.askForOnDeviceConfirmation,
-        chainCode
+  async getMultipleAccounts(indexOffset, accountsNo) {
+    let eth = null;
+    try {
+      if (!this.isU2FSupported) {
+        return Promise.reject(new Error(NOT_SUPPORTED_ERROR_MSG));
```
Comment (krzkaczor) [inline src/LedgerWallet.js]: just throw error here. it will automatically reject promise (move it out of try probably)

File: src/LedgerWallet.js [hunk 2]
```diff
@@ -158,109 +148,139 @@ class LedgerWallet {
...
+          pathComponents.basePath + (pathComponents.index + i).toString();
+        // eslint-disable-next-line no-await-in-loop
+        const address = await eth.getAddress_async(
+          path,
+          this.askForOnDeviceConfirmation,
+          chainCode
+        );
+        addresses[path] = address.address;
+      }
+      return Promise.resolve(addresses);
```
Comment (krzkaczor) [inline src/LedgerWallet.js]: you use await async here. Just return

File: src/LedgerWallet.js [hunk 3]
```diff
@@ -158,109 +148,139 @@ class LedgerWallet {
...
+    } catch (e) {
+      return Promise.reject(e);
```
Comment (krzkaczor) [inline src/LedgerWallet.js]: i think you can ommit catch here entirely: https://stackoverflow.com/a/5764505/580181 if not please rethrow error here with `throw e` (but i am pretty sure you dont need it)

File: src/LedgerWallet.js [hunk 4]
```diff
@@ -158,109 +148,139 @@ class LedgerWallet {
...
+    } finally {
+      if (eth !== null) {
+        // This is fishy but currently ledger library always returns empty
+        // resolved promise when closing connection so there is no point in
+        // doing anything with returned Promise.
+        // noinspection JSIgnoredPromiseFromCall
+        this.closeLedgerConnection(eth);
```
Comment (krzkaczor) [inline src/LedgerWallet.js]: awaiting on this promise makes sense - it can help avoid race conditions etc

File: src/LedgerWallet.js [hunk 5]
```diff
@@ -158,109 +148,139 @@ class LedgerWallet {
...
+    return { basePath: matchResult[1], index: parseInt(matchResult[2], 10) };
+  }
 
-    // Encode as hex-rlp for Ledger
-    const hex = tx.serialize().toString("hex");
+  async signTransactionAsync(txData) {
+    let eth = null;
+    try {
+      if (!this.isU2FSupported) {
+        return Promise.reject(new Error(NOT_SUPPORTED_ERROR_MSG));
```
Comment (krzkaczor) [inline src/LedgerWallet.js]: throw new

File: src/LedgerWallet.js [hunk 6]
```diff
@@ -158,109 +148,139 @@ class LedgerWallet {
...
-          cleanupCallback(
+      // Store signature in transaction
+      tx.v = Buffer.from(result.v, "hex");
+      tx.r = Buffer.from(result.r, "hex");
+      tx.s = Buffer.from(result.s, "hex");
+
+      // EIP155: v should be chain_id * 2 + {35, 36}
+      const signedChainId = Math.floor((tx.v[0] - 35) / 2);
+      if (signedChainId !== chainId) {
+        return Promise.reject(
```
Comment (krzkaczor) [inline src/LedgerWallet.js]: throw

File: src/LedgerWallet.js [hunk 7]
```diff
@@ -158,109 +148,139 @@ class LedgerWallet {
...
+      }
 
-        // Return the signed raw transaction
-        const rawTx = `0x${tx.serialize().toString("hex")}`;
-        cleanupCallback(undefined, rawTx);
-      })
-      .catch(error => cleanupCallback(error));
+      // Return the signed raw transaction
+      const rawTx = `0x${tx.serialize().toString("hex")}`;
+      return Promise.resolve(rawTx);
```
Comment (krzkaczor) [inline src/LedgerWallet.js]: just return

File: src/LedgerWallet.js [hunk 8]
```diff
@@ -158,109 +148,139 @@ class LedgerWallet {
...
+    } catch (e) {
+      return Promise.reject(e);
```
Comment (krzkaczor) [inline src/LedgerWallet.js]: remove

File: src/LedgerWallet.js [hunk 9]
```diff
@@ -158,109 +148,139 @@ class LedgerWallet {
...
+    } finally {
+      if (eth !== null) {
+        // This is fishy but currently ledger library always returns empty
+        // resolved promise when closing connection so there is no point in
+        // doing anything with returned Promise.
+        // noinspection JSIgnoredPromiseFromCall
+        this.closeLedgerConnection(eth);
```
Comment (krzkaczor) [inline src/LedgerWallet.js]: await

File: src/LedgerWallet.js [hunk 10]
```diff
@@ -249,29 +234,35 @@ class LedgerWallet {
...
-      return Promise.resolve(rawTx);
-    } catch (e) {
-      return Promise.reject(e);
+      return `0x${tx.serialize().toString("hex")}`;
     } finally {
       if (eth !== null) {
         // This is fishy but currently ledger library always returns empty
         // resolved promise when closing connection so there is no point in
         // doing anything with returned Promise.
         // noinspection JSIgnoredPromiseFromCall
```
Comment (krzkaczor) [inline src/LedgerWallet.js]: are these noinspection lines still valid? you have await on the result. And on the other hand I would not pollute our codebase with InteliJ products specific comments. You shall silent it somehow in you project config. We should stick to eslint for linting.

Comment (banciur) [reply src/LedgerWallet.js]: You are totally right now its not required but anyway it should never get there.

**Pull request merged**